src/osd/OSD.cc

```diff
@@ -4217,14 +4218,40 @@ void OSD::resume_creating_pg()
       pg = pending_creates_from_osd.erase(pg);
       spare_pgs--;
     }
+    have_pending_creates = (pending_creates_from_mon > 0 ||
+			    !pending_creates_from_osd.empty());
   }
+
+  bool do_renew_subs = false;
   if (do_sub_pg_creates) {
     if (monc->sub_want("osd_pg_creates", last_pg_create_epoch, 0)) {
       dout(4) << __func__ << ": resolicit pg creates from mon since "
 	      << last_pg_create_epoch << dendl;
-      monc->renew_subs();
+      do_renew_subs = true;
+    }
+  }
+  version_t start = osdmap->get_epoch() + 1;
+  if (have_pending_creates) {
+    // don't miss any new osdmap deleting PGs
+    if (monc->sub_want("osdmap", start, 0)) {
+      dout(4) << __func__ << ": resolicit osdmap from mon since "
+	      << start << dendl;
+      do_renew_subs = true;
+    }
+  } else if (pgtemp || do_sub_pg_creates) {
+    // no need to subscribe the osdmap continuously anymore
+    // once the pgtemp and/or mon_subscribe(pg_creates) is sent
+    if (monc->sub_want_increment("osdmap", start, CEPH_SUBSCRIBE_ONETIME)) {
+      dout(4) << __func__ << ": re-subscribe osdmap(onetime) since"
+	      << start << dendl;
+      do_renew_subs = true;
```

Comment:
Once we have pending creates and a full subscription, how does it stop the subscription after the pg gets created?

Reply:
@liewegas that's what https://github.com/ceph/ceph/pull/18916/files#diff-fa6c2eba8356ae1442d1bf749beacfdfR4244 is for.

Reply:
what if !(pgtemp || do_sub_pg_creates) thought?

Reply:
@liewegas let me rephrase the code:
```c++
if (have_pending_creates) {
  subscribe(osdmap, continuously);
}
if (!have_pending_create) {
  if (pgtemp || do_sub_pg_creates) {
     /// the reason why we don't have pending creates is that
     /// we are about to create the pending pgs, and that means
     /// we've being subscribing the osdmap continuously, so 
     subscribe(osdmap, onetime); 
  } else {
     /// awesome, no pending creates now, and no pending creates **before** calling resume_creating_pg()
  }
}
```